### PR TITLE
fix: Support Dark theme based on the web browser's theme

### LINKED ISSUE #16: Support Dark theme based on the web browser's theme
Now the project doesn't switch to dark theme when the browser's theme is set to dark, we have to make it compatible for dark theme. 
To better test this, use edge browser and switch the theme between light and dark.

### PR DESCRIPTION
**Pull request by SankalpHaritash21**
## Summary
Adding Dark Theme feature.

## Description
Updating Dark Theme According to Figma design

## Images
Add-page:

add-page

details-page:
detail-page

Home-page:
home 1

home 2

## Issue(s) Addressed
- Closed #16

## Review Checklist
Please make sure to go through this checklist before submitting your review.

- [X] Make sure the PR title is of format **`tag: Issue Title`**
   - tag should be one of `fix`,`feat`, `docs`, `chore`, `refactor`
   - Ex: `feat: Added responsive design for the home page`
- [X] Ensure the code follows our coding standards and guidelines
   - Ex:  naming conventions should be of form `file-name.ts` not `fileName.ts`
- [X] Ensure that you have staged only the required commits and raised the PR and not included all the files which would be otherwise required for local setup
- [X] Check if the PR is rebased to the latest main/development branch.
- [X] Make sure there are no unnecessary files or debugging artifacts in the PR.

### CODE REVIEW

File: frontend/src/index.css [hunk 1]
```diff
@@ -1,3 +1,9 @@
 @tailwind base;
 @tailwind components;
 @tailwind utilities;
+
+@layer utilities {
+  .dark {
```
Comment (krishnaacharyaa) [inline frontend/src/index.css:6]: What is the use case of this ? And why is this added ?

File: frontend/src/pages/add-blog.tsx [hunk 2]
```diff
@@ -89,95 +89,99 @@ function AddBlog() {
...
-          <textarea
-            name="description"
-            placeholder="Description"
-            className="w-full p-2 rounded-lg bg-gray-50 placeholder:text-gray-800 "
-            value={formData.description}
-            onChange={handleInputChange}
-          />
-        </div>
+        <form onSubmit={handleSubmit}>
+          <div className="mb-4">
```
Comment (krishnaacharyaa) [inline frontend/src/pages/add-blog.tsx:102]: Please check the formatting, though the files are not changed why are we seeing the changes

File: frontend/src/pages/add-blog.tsx [hunk 3]
```diff
@@ -89,95 +89,99 @@ function AddBlog() {
...
+          </div>
 
-        <div className="mb-4 flex items-center flex-col md:flex-row ">
-          <label className="block font-semibold mb-2 w-full md:w-fit  md:mr-8">Categories</label>
-          <div className="flex flex-wrap gap-2">
-            {CATEGORIES.map((category) => (
-              <span
-                key={category}
-                className={`cursor-pointer
+          <div className="mb-4 flex items-center flex-col md:flex-row ">
```
Comment (krishnaacharyaa) [inline frontend/src/pages/add-blog.tsx:142]: same here

File: frontend/src/pages/add-blog.tsx [hunk 4]
```diff
@@ -89,95 +89,99 @@ function AddBlog() {
...
+            <label className="block font-semibold mb-2 w-full md:w-fit  md:mr-8">Categories</label>
+            <div className="flex flex-wrap gap-2 dark:text-black">
+              {CATEGORIES.map((category) => (
+                <span
+                  key={category}
+                  className={`cursor-pointer
 									${
                     formData.categories.includes(category)
```
Comment (krishnaacharyaa) [inline frontend/src/pages/add-blog.tsx:150]: Check the formatting please

File: frontend/src/pages/add-blog.tsx [hunk 5]
```diff
@@ -89,95 +89,99 @@ function AddBlog() {
...
-          type="submit"
-          className="bg-black text-white flex items-center justify-center text-base p-2 rounded-lg hover:bg-gray-800 w-full md:w-fit"
-        >
-          Create Blog
-        </button>
-      </form>
+          <button
+            type="submit"
+            className=" dark:border-2 dark:border-white dark:bg-orange-400 bg-black text-white flex items-center justify-center text-base md:text-lg lg:text-xl p-2 rounded-lg hover:bg-gray-800 w-full md:w-fit"
+          >
```
Comment (krishnaacharyaa) [inline frontend/src/pages/add-blog.tsx:180]: For dark mode we can use white button and black text, would look nice than orange

File: frontend/src/pages/details-page.tsx [hunk 6]
```diff
@@ -8,36 +8,38 @@ const DetailsPage = () => {
...
   return (
-    <div className="w-full font-[Poppins] ">
-      <div className="relative flex flex-col">
-        <img
-          src={post.imageLink}
-          alt={post.title}
-          className="h-[360px] md:h-[460px] object-cover w-full"
-        />
+    <div className="dark:dark w-full min-h-screen">
+      <div className="w-full font-[Poppins]">
```
Comment (krishnaacharyaa) [inline frontend/src/pages/details-page.tsx:12]: Okay if you have used dark:dark here why not all the places the same? some has dark:text-white this one has dark:dark, i dont see consistency and the need of this extra confusing custom dark utility class. kindly confirm once

File: frontend/src/pages/home-page.tsx [hunk 7]
```diff
@@ -20,8 +20,8 @@ function HomePage() {
   }, []);
   const navigate = useNavigate();
   return (
-    <div className="font-[Poppins] w-full">
-      <div className="relative h-[460px] bg-cover mt-[-8px] bg-fixed bg-center bg-[url('https://i.ibb.co/Pm4ScWL/wanderlust-bg.webp')]">
+    <div className="font-[Poppins] w-full dark:dark">
+      <div className="relative h-[460px] bg-cover mt-[-8px] bg-fixed bg-center bg-[url('https://get.wallhere.com/photo/landscape-sea-beach-hotel-swimming-pool-resort-tropical-lagoon-Caribbean-vacation-estate-leisure-villa-225624.jpg')]">
```
Comment (krishnaacharyaa) [inline frontend/src/pages/home-page.tsx:24]: Don't revert the url, let it be same as before

**Review by krishnaacharyaa — CHANGES_REQUESTED** [review]
Please go through the comments

File: frontend/src/pages/home-page.tsx (hunk 7 above)
Comment (SankalpHaritash21) [reply frontend/src/pages/home-page.tsx:24]: sir I han't reverted the URl.I

File: frontend/src/pages/details-page.tsx (hunk 6 above)
Comment (SankalpHaritash21) [reply frontend/src/pages/details-page.tsx:12]: sir some places we only need text as black. and some places we need background and text both as black

File: frontend/src/index.css (hunk 1 above)
Comment (SankalpHaritash21) [reply frontend/src/index.css:6]: sir should I keep dark or should I remove it

**Comment by krishnaacharyaa** [comment]
For the cards also consider applying dark theme in the home page like below, get the colors of the image using colour picker and match it, card should look something like this background in dark mode instead of white
image

File: frontend/src/index.css (hunk 1 above)
Comment (krishnaacharyaa) [reply frontend/src/index.css:6]: I am asking the use case, I am trying to understand why have you added that, once you say, I can tell you if it is correct or if there is any other better way

File: frontend/src/pages/details-page.tsx (hunk 6 above)
Comment (krishnaacharyaa) [reply frontend/src/pages/details-page.tsx:12]: what is the case where both background and text are black ? I didn't get you, and dark:dark only for this place I don't think it is efficient you can directly replace class names inside the dark here, it is not good dev experience

File: frontend/src/pages/home-page.tsx (hunk 7 above)
Comment (krishnaacharyaa) [reply frontend/src/pages/home-page.tsx:24]: At least I can see it as changed

File: frontend/src/pages/add-blog.tsx (hunk 5 above)
Comment (krishnaacharyaa) [inline frontend/src/pages/add-blog.tsx:180]: Lets have text-base for all the scenario, it is the latest changes done by me, please revert it back

File: frontend/src/index.css (hunk 1 above)
Comment (SankalpHaritash21) [reply frontend/src/index.css:6]: Sir I added .dark at global level it will automatically gives bg as black and text as white. This will help in adding dark mode at globel level.

File: frontend/src/pages/details-page.tsx (hunk 6 above)
Comment (SankalpHaritash21) [reply frontend/src/pages/details-page.tsx:12]: ok sir I remove it

File: frontend/src/pages/add-blog.tsx (hunk 5 above)
Comment (SankalpHaritash21) [reply frontend/src/pages/add-blog.tsx:180]: sir text-base is looking too small for laptop mode

Comment (krishnaacharyaa) [reply frontend/src/pages/add-blog.tsx:180]: No, 16px is fine for even larger screen, if you make the text larger, you would have to also have a responsive padding for it , else it won't follow the UI guidelines

File: frontend/src/index.css (hunk 1 above)
Comment (krishnaacharyaa) [reply frontend/src/index.css:6]: Its not convincing @SankalpHaritash21, you need to either convince why using this is best if you feel it is, or else we will remove it as it is used in only place if you can make it at global level, then you have to use it globally, i see it being used only at one place, so doesn't server the purpose of being global afterall

File: frontend/src/pages/add-blog.tsx (hunk 5 above)
Comment (SankalpHaritash21) [reply frontend/src/pages/add-blog.tsx:180]: Sir where can i find ui guidelines

Comment (SankalpHaritash21) [reply frontend/src/pages/add-blog.tsx:180]: Sir can i add little bit shadow

Comment (SankalpHaritash21) [reply frontend/src/pages/add-blog.tsx:180]: At bottom

**Comment by krishnaacharyaa** [comment]
> For the cards also consider applying dark theme in the home page like below, get the colors of the image using colour picker and match it, card should look something like this background in dark mode instead of white image

Kindly do this changes in the home page all the cards and also in the label of the add blog page, lets use the background of this color and also background of card similar to this.

**Comment by krishnaacharyaa** [comment]
Please dont use pitch black anywhere i.e bg-black X use the above background color and card color from the shared image

File: frontend/src/index.css (hunk 1 above)
Comment (SankalpHaritash21) [reply frontend/src/index.css:6]: Ok sir I will remove it and add it simply to  classes

**Comment by SankalpHaritash21** [comment]
Sir can I keet bg as black for home and card bg will be light black as shown in image

File: frontend/src/pages/add-blog.tsx (hunk 5 above)
Comment (krishnaacharyaa) [reply frontend/src/pages/add-blog.tsx:180]: material 3 guidelines from google

Comment (krishnaacharyaa) [reply frontend/src/pages/add-blog.tsx:180]: It should be subtle, dont make it 3D

**Comment by krishnaacharyaa** [comment]
> Sir can I keet bg as black for home and card bg will be light black as shown in image

For even the background use image's background. 
"Sir can I keet bg as black for home", for this i already answered dont use pitch black

**Comment by SankalpHaritash21** [comment]
sir I want to ask have I updated my PR correctly. plz tell.

**Comment by krishnaacharyaa** [comment]
You have not used the said colors 🙃, let me create figma design for you

**Comment by SankalpHaritash21** [comment]
is can u plz give be correct color for background i have used bg-slate-700 and 500 and text as white

**Comment by krishnaacharyaa** [comment]
I dont think @SankalpHaritash21, kindly compare the image i have provided with the images you have updated, they are very different from each other, have a close look.
I am know working on the dark mode, give me sometime, i'll get back to you.

**Comment by krishnaacharyaa** [comment]
I have updated the figma, please go through that, take your sweet time and make sure proper padding rounded corners and consistency and best practices is followed when using dark theme

**Comment by SankalpHaritash21** [comment]
Sir in figma design is background black or gray plz confirm.
Sir padding will remain same for light and dark mode or I have to update.

**Comment by krishnaacharyaa** [comment]
@SankalpHaritash21 , do we have any updates on this, kindly let us know here :)

**Comment by SankalpHaritash21** [comment]
Yes, @krishnaacharyaa  currently I am working on this approach https://youtu.be/MAtaT8BZEAo?si=AFaK_x-M15OKLsNI . I am trying to optimize function call.

**Comment by krishnaacharyaa** [comment]
Thats awesome  @SankalpHaritash21, Thank you for your update, kindly let us know here may be once in 2 days about the progress that you have made, so that i can guide/suggest you if needbe.
The extra efforts of you are really appreciated :)

**Comment by SankalpHaritash21** [comment]
Good evening sir, Sir I am Facing too much merge conflict while rebasing can u plz suggest some solution.

**Comment by krishnaacharyaa** [comment]
@SankalpHaritash21, yeah sankalp there are too many changes we underwent recently, there are 2 things you can do
1. Rebase your branch, accept the upstream for everything and then commit and push it , then you can add your changes in the next commit.
2. Create new branch from your main branch once your main is in sync with upstream's main (rebase your master with this project) 5th step in the [contributing guidelines](https://github.com/krishnaacharyaa/wanderlust/blob/bd3db1200378fe08dab7cfedca5d0e341b7a187b/.github/CONTRIBUTING.md) and then cherry-pick all of your commits one by one, if there are any merge conflict try resolving.

If you want to get off of this merge conflict all together rebase your main to upstream i.e this project and then create new branch and then manual make changes again, referring your commits.

See which would be helpful for you. I guess you can go with last option is better though it is time consuming. But if you want to master git , i would recommend you try the 1st or 2nd one ... which involves solving the merge conflicts

**Comment by SankalpHaritash21** [comment]
sir I tryed but not able to resolve I want to learn git plz help

**Comment by krishnaacharyaa** [comment]
The error you are facing is because you haven't rebased your branch and resolved @SankalpHaritash21, can you patiently go through https://github.com/krishnaacharyaa/wanderlust/pull/43#issuecomment-1798469636, it answers all of your problems, kindly do some research, you'll learn and then it will get solved.

If merge conflict is headache to you and you just want to get done with your PR, you can consider creating new branch from main branch after you do git pull upstream main. And then you can start making the changes from the beginning, may be a long process, but you wont be worked up :)

**Comment by SankalpHaritash21** [comment]
sir can I take some time. I will try to resolve conflict

File: frontend/src/pages/home-page.tsx [hunk 8]
```diff
@@ -20,9 +20,8 @@ function HomePage() {
   }, []);
   const navigate = useNavigate();
   return (
-    <div className="font-[Poppins] w-full">
-      {/* <img src={bg} alt="" /> */}
-      <div style={{backgroundImage: `url(${bg})`}} className="relative h-[460px] bg-cover mt-[-8px] bg-fixed bg-center" >
+    <div className="font-[Poppins] w-full dark:bg-dark-bg dark:text-dark-text">
+      <div className="relative h-[460px] bg-cover mt-[-8px] bg-fixed bg-center bg-[url('https://i.ibb.co/Pm4ScWL/wanderlust-bg.webp')]">
```
Comment (krishnaacharyaa) [inline frontend/src/pages/home-page.tsx:24]: Why are we complicating to use dark:bg-dark-bg ? I assume this is related to tailwind.config.js , this seems confusing , why can't we use directly dark:bg-dark ? from index.css, May i know what is the intension behind this workaround, where it could have been normal

File: frontend/tailwind.config.js [hunk 9]
```diff
@@ -1,8 +1,18 @@
...
+  darkMode: 'media',
+  theme: {
+    extend: {
+      colors: {
+        dark: {
+          bg: 'var(--bg-dark)',
+          field: 'var(--bg-dark-field)',
+          card: 'var(--bg-dark-card)',
+          text: 'var(--text-dark)',
+        },
```
Comment (krishnaacharyaa) [inline frontend/tailwind.config.js:13]: What is the use of using this ? Please justify, so that we know if it actually required

File: frontend/src/pages/details-page.tsx [hunk 10]
```diff
@@ -1,43 +1,45 @@
...
   return (
-    <div className="w-full font-[Poppins] ">
-      <div className="relative flex flex-col">
-        <img
-          src={post.imageLink}
-          alt={post.title}
-          className="h-[360px] md:h-[460px] object-cover w-full"
-        />
+    <div className="dark:bg-dark-bg dark:text-dark-text w-full min-h-screen">
+      <div className="w-full font-[Poppins]">
```
Comment (krishnaacharyaa) [inline frontend/src/pages/details-page.tsx:12]: same here with dark:text-dark-text

File: frontend/src/pages/add-blog.tsx [hunk 11]
```diff
@@ -89,95 +89,105 @@ function AddBlog() {
...
-    <div className="font-[Poppins] p-4 px-16 bg-white ">
-      <div className="flex justify-start items-center mb-4">
-        <div className=" text-black cursor-pointer w-fit  text-base md:text-lg lg:text-2xl">
-        <img src={blackleftarrow}  style={{height:20, width:40}}onClick={() => navigate(-1)} />
-        </div>
-        <h2 className="text-sm md:text-lg  lg:text-2xl font-bold ml-4">Create Post</h2>
-      </div>
+    <div className=" w-full min-h-screen bg-dark-bg">
+      <div className="w-full min-h-screen ">
+        <div className="font-[Poppins] p-4 px-16 bg-white dark:bg-dark-bg ">
```
Comment (krishnaacharyaa) [inline frontend/src/pages/add-blog.tsx:94]: Do we actually need this extra divs ? can't we simplify this

**Review by krishnaacharyaa — CHANGES_REQUESTED** [review]
Please look into the given comments

File: frontend/src/pages/home-page.tsx (hunk 8 above)
Comment (SankalpHaritash21) [reply frontend/src/pages/home-page.tsx:24]: sir when there are 2 color then it create conflict. And if we want to give opacity it will not take directly. Then I decided to declare color as variable in config file. @krishnaacharyaa  dark:bg-dark-bg it seems to be complicated but it will be effective when we have more custom colors

File: frontend/tailwind.config.js (hunk 9 above)
Comment (SankalpHaritash21) [reply frontend/tailwind.config.js:13]: sir this help in to give variable or definition to color

File: frontend/src/pages/add-blog.tsx (hunk 11 above)
Comment (SankalpHaritash21) [reply frontend/src/pages/add-blog.tsx:94]: I will recheck and I will try to reduse div

File: frontend/src/pages/home-page.tsx (hunk 8 above)
Comment (krishnaacharyaa) [reply frontend/src/pages/home-page.tsx:24]: I respectfully disagree with your thoughts @SankalpHaritash21, dark:bg-dark-bg would be more complicated when we have more custom colors this will be all around the place, not a good dev experience, we need to avoid using this

Comment (krishnaacharyaa) [reply frontend/src/pages/home-page.tsx:24]: @SankalpHaritash21, Okay after thoroughly going through the code, and referring some docs i figured you are trying to dynamically get the colors, thats good, but just one tip, "I got confused till i saw the code myself" what does it mean ? New contributors will be clueless of how this is applied and where is this coming from, so what do you need to do ? 
**Simplify the complexity**, how do you do that ? By adding good naming conventions.

`bg-dark-bg` can we make it to `bg-dark-background` though it is long but it is now much clear, otherwise you get confused with the tailwind's pre determined classes, isn't it . Same like `text-dark-text` dark-text what ? dark-text color ? font ? size ? correct ? , so make it to `text-dark-text-color`

Comment (krishnaacharyaa) [reply frontend/src/pages/home-page.tsx:24]: Just one thing, please make if the vs code tailwind extension will pick this up ? otherwise it will be bit of a pain

File: frontend/tailwind.config.js (hunk 9 above)
Comment (krishnaacharyaa) [reply frontend/tailwind.config.js:13]: @SankalpHaritash21, Please work on the variable names as mentioned above, even with css variable
`card: 'var(--bg-dark-card)'` it is saying color of `bg-dark-card` to `text` and you are using this for text color 
chaning it to something like `dark-card-text-color` would be less confusing and easy to interrupt.

File: frontend/src/pages/home-page.tsx (hunk 8 above)
Comment (SankalpHaritash21) [reply frontend/src/pages/home-page.tsx:24]: Yes @krishnaacharyaa  tailwind intellisence in vs code is detecting these color.
I will change the name and will give them a standard name.

Comment (SankalpHaritash21) [reply frontend/src/pages/home-page.tsx:24]: sir I will take some time and will try to come up with better solution this time.
@krishnaacharyaa  can I adjust little bit padding in home page or it will be another issue.
And plz let me know is design good. Plz check Images in PR and let me know

Comment (krishnaacharyaa) [reply frontend/src/pages/home-page.tsx:24]: Yeah you can add 16px padding on the right and top of featured post card :)

File: frontend/src/pages/add-blog.tsx (hunk 11 above)
Comment (SankalpHaritash21) [reply frontend/src/pages/add-blog.tsx:94]: @krishnaacharyaa  variable name changed, add blog and detail page completed . only home page remaning will complete.

Comment (krishnaacharyaa) [reply frontend/src/pages/add-blog.tsx:94]: Sure @SankalpHaritash21 Thank you for the update

**Comment by SankalpHaritash21** [comment]
@krishnaacharyaa  sir I am trying to resolve conflicts. Do I have to review all updated codes and update them accordingly.
@krishnaacharyaa sorry for taking too much time in single task.
